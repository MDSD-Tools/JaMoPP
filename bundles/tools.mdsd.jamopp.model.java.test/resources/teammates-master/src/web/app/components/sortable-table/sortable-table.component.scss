.sortable-header {
  cursor: pointer;

  i {
    margin-left: .3em;
  }
}

.margin-bottom-20px {
  margin-bottom: 20px;
}
