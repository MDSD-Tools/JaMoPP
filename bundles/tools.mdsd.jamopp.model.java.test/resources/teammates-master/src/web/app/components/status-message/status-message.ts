/**
 * Represents a status message and a severity color.
 *
 * The color is adapted from Bootstrap's color scheme.
 */
export interface StatusMessage {
  message: string;
  color: string;
}
