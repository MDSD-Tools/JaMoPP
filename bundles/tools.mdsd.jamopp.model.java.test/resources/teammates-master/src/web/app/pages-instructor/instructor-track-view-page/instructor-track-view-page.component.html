<h1>Student Result View Tracking</h1>
<p>This page allows you to track if the session results were viewed by students. The search will go only up to the date the session was published. The earliest date you can search for is <b>{{this.LOGS_RETENTION_PERIOD_IN_DAYS}} days</b> before today.</p>
<hr/>

<div *tmIsLoading="isLoading" class="card bg-form no-border">
  <div class="card-body">
    <div class="row form-group">
      <div class="col-5">
        <div class="col-12">
          <label for="course-id-dropdown" class="font-weight-bold">Course ID</label>
        </div>
        <div class="col-12">
          <select id="course-id-dropdown" class="form-control col-12" [(ngModel)]="formModel.courseId">
            <option value="" disabled selected hidden>Select Course ID</option>
            <option *ngFor="let course of courses">
              {{course.courseId}}
            </option>
          </select>
        </div>
      </div>
      <div class="col-5">
        <div class="col-12">
          <label for="session-dropdown" class="font-weight-bold">Feedback Session Name</label>
        </div>
        <div class="col-12">
          <select id="session-dropdown" class="form-control col-md-12" [(ngModel)]="formModel.feedbackSessionName">
            <option *ngFor="let session of courseToFeedbackSession[formModel.courseId]" [value]="session.feedbackSessionName">
              {{session.feedbackSessionName}}
            </option>
          </select>
        </div>
      </div>
      <div class="col-1 mt-4">
        <button id="view-button" class="float-right btn btn-primary" (click)="getFeedbackSession()" [disabled]="!formModel.courseId || !formModel.feedbackSessionName || isSearching">View</button>
      </div>
    </div>
  </div>
</div>

<div *tmIsLoading="isSearching">
  <div *ngIf="hasResult">
    <p class="font-weight-bold text-break">{{searchResult.feedbackSessionName}} in {{searchResult.courseId}} (published on: {{searchResult.publishedDate}})</p>
    <tm-sortable-table [rows]="searchResult.logRowsData" [columns]="searchResult.logColumnsData" [initialSortBy]="SortBy.RESULT_VIEW_STATUS"></tm-sortable-table>
  </div>
</div>
