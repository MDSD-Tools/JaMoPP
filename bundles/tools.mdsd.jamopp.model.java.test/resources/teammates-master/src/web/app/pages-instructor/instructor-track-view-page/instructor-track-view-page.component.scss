.bg-form {
  background-color: #EAEFF5;
}

.no-border {
  border: 0;
}
