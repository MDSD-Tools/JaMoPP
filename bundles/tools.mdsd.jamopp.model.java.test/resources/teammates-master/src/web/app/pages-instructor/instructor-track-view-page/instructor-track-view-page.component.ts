import { Component, OnInit } from '@angular/core';
import { NgbDateParserFormatter, NgbModalRef } from '@ng-bootstrap/ng-bootstrap';
import { concatMap, finalize, mergeAll } from 'rxjs/operators';
import { CourseService } from '../../../services/course.service';
import { FeedbackSessionsService } from '../../../services/feedback-sessions.service';
import { LogService } from '../../../services/log.service';
import { SimpleModalService } from '../../../services/simple-modal.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { StudentService } from '../../../services/student.service';
import { TimezoneService } from '../../../services/timezone.service';
import { ApiConst } from '../../../types/api-const';
import {
  Course,
  Courses,
  FeedbackSession,
  FeedbackSessionLog,
  FeedbackSessionLogEntry,
  FeedbackSessionLogs,
  FeedbackSessionLogType,
  FeedbackSessionPublishStatus,
  FeedbackSessions,
  Student,
  Students,
} from '../../../types/api-output';
import { Intent } from '../../../types/api-request';
import { SortBy } from '../../../types/sort-properties';
import { DateFormat } from '../../components/datepicker/datepicker.component';
import { SessionEditFormDatePickerFormatter } from '../../components/session-edit-form/session-edit-form-datepicker-formatter';
import { SimpleModalType } from '../../components/simple-modal/simple-modal-type';
import { ColumnData, SortableTableCellData } from '../../components/sortable-table/sortable-table.component';
import { TimeFormat } from '../../components/timepicker/timepicker.component';
import { ErrorMessageOutput } from '../../error-message-output';

/**
 * Model for searching of logs
 */
interface SearchLogsFormModel {
  courseId: string;
  feedbackSessionName: string;
}

/**
 * Model for displaying of feedback session logs
 */
interface FeedbackSessionLogModel {
  courseId: string;
  feedbackSessionName: string;
  publishedDate: string;
  logColumnsData: ColumnData[];
  logRowsData: SortableTableCellData[][];
}

/**
 * Component for tracking student results view
 */
@Component({
  selector: 'tm-instructor-track-view-page',
  templateUrl: './instructor-track-view-page.component.html',
  providers: [{ provide: NgbDateParserFormatter, useClass: SessionEditFormDatePickerFormatter }],
  styleUrls: ['./instructor-track-view-page.component.scss'],
})
export class InstructorTrackViewPageComponent implements OnInit {
  LOGS_DATE_TIME_FORMAT: string = 'ddd, DD MMM YYYY hh:mm:ss A';
  LOGS_RETENTION_PERIOD_IN_DAYS: number = ApiConst.LOGS_RETENTION_PERIOD;
  LOGS_RETENTION_PERIOD_IN_MILLISECONDS: number = this.LOGS_RETENTION_PERIOD_IN_DAYS * 24 * 60 * 60 * 1000;
  ONE_MINUTE_IN_MILLISECONDS: number = 60 * 1000;

  // enum
  SortBy: typeof SortBy = SortBy;

  formModel: SearchLogsFormModel = {
    courseId: '',
    feedbackSessionName: '',
  };
  courses: Course[] = [];
  courseToFeedbackSession: Record<string, FeedbackSession[]> = {};
  searchResult: FeedbackSessionLogModel = {
    courseId: '',
    feedbackSessionName: '',
    publishedDate: '',
    logColumnsData: [],
    logRowsData: [],
  };
  students: Student[] = [];
  studentToLog: Record<string, FeedbackSessionLogEntry> = {};
  isLoading: boolean = true;
  isSearching: boolean = false;
  hasResult: boolean = false;
  publishedTime: number = 0;
  notViewedSince: number = 0;
  logsDateFrom: DateFormat = { year: 0, month: 0, day: 0 };
  logsTimeFrom: TimeFormat = { hour: 23, minute: 59 };
  logsDateTo: DateFormat = { year: 0, month: 0, day: 0 };
  logsTimeTo: TimeFormat = { hour: 23, minute: 59 };

  constructor(private courseService: CourseService,
    private feedbackSessionsService: FeedbackSessionsService,
    private logsService: LogService,
    private simpleModalService: SimpleModalService,
    private statusMessageService: StatusMessageService,
    private studentService: StudentService,
    private timezoneService: TimezoneService) { }

  ngOnInit(): void {
    this.loadData();
  }

  /**
   * Load all courses that the instructor have
   */
  private loadData(): void {
    this.courseService
      .getAllCoursesAsInstructor('active')
      .pipe(
          concatMap((courses: Courses) => courses.courses.map((course: Course) => {
            this.courses.push(course);
            return this.feedbackSessionsService.getFeedbackSessionsForInstructor(course.courseId);
          })),
          mergeAll(),
          finalize(() => this.isLoading = false))
      .subscribe(((feedbackSessions: FeedbackSessions) => {
        if (feedbackSessions.feedbackSessions.length > 0) {
          this.courseToFeedbackSession[feedbackSessions.feedbackSessions[0].courseId]
            = [...feedbackSessions.feedbackSessions];
        }
      }),
      (e: ErrorMessageOutput) => this.statusMessageService.showErrorToast(e.error.message));
  }

  /**
   * Get the selected feedback session
   */
  getFeedbackSession(): void {
    this.isSearching = true;

    // Get the publish status and time of feedback session
    this.feedbackSessionsService
      .getFeedbackSession({
        courseId: this.formModel.courseId,
        feedbackSessionName: this.formModel.feedbackSessionName,
        intent: Intent.INSTRUCTOR_RESULT,
      })
      .subscribe((feedbackSession: FeedbackSession) => {
        this.publishedTime = feedbackSession.resultVisibleFromTimestamp || this.publishedTime;

        // Feedback session is not published, do not need to search.
        if (feedbackSession.publishStatus !== FeedbackSessionPublishStatus.PUBLISHED) {
          this.isSearching = false;
          this.statusMessageService.showErrorToast('This feedback session is not published');
          return;
        }

        const earliestSearchTime: number = Date.now() - this.LOGS_RETENTION_PERIOD_IN_MILLISECONDS;
        // Feedback session is published more than 30 days ago, open a dialog.
        if (this.publishedTime < earliestSearchTime) {
          this.openModal();
          return;
        }

        // Published time of the feedback session is less than 30 days ago, update the form model to
        // search only up till the publish date.
        const publishedDate: Date = new Date(this.publishedTime);
        this.notViewedSince = publishedDate.getTime();
        this.logsDateFrom = {
          year: publishedDate.getFullYear(),
          month: publishedDate.getMonth() + 1,
          day: publishedDate.getDate(),
        };
        this.logsTimeFrom = {
          hour: publishedDate.getHours(),
          minute: publishedDate.getMinutes(),
        };
        this.search();
      });
  }

  /**
   * Search for logs of student activity
   */
  private search(): void {
    this.students = [];
    this.studentToLog = {};

    const today: Date = new Date();
    this.logsDateTo = {
      year: today.getFullYear(),
      month: today.getMonth() + 1,
      day: today.getDate(),
    };

    const logsDateFrom: number = this.timezoneService.resolveLocalDateTime(this.logsDateFrom, this.logsTimeFrom);
    const logsDateTo: number = this.timezoneService.resolveLocalDateTime(this.logsDateTo, this.logsTimeTo);

    this.logsService.searchFeedbackSessionLog({
      courseId: this.formModel.courseId,
      searchFrom: logsDateFrom.toString(),
      searchUntil: logsDateTo.toString(),
      sessionName: this.formModel.feedbackSessionName,
    }).pipe(
        finalize(() => {
          this.isSearching = false;
          this.hasResult = true;
        }),
    ).subscribe((logs: FeedbackSessionLogs) => {
      this.studentService
          .getStudentsFromCourse({ courseId: this.formModel.courseId })
          .subscribe((students: Students) => {
            this.students.push(...students.students);

            const targetFeedbackSessionLog: FeedbackSessionLog | undefined = logs.feedbackSessionLogs
                .find((fsLog: FeedbackSessionLog) =>
                    fsLog.feedbackSessionData.feedbackSessionName === this.formModel.feedbackSessionName);
            if (!targetFeedbackSessionLog) {
              return;
            }

            targetFeedbackSessionLog.feedbackSessionLogEntries
                .filter((entry: FeedbackSessionLogEntry) =>
                    entry.feedbackSessionLogType.toString() as keyof typeof FeedbackSessionLogType
                    === 'VIEW_RESULT')
                .filter((entry: FeedbackSessionLogEntry) =>
                    !(entry.studentData.email in this.studentToLog)
                    || this.studentToLog[entry.studentData.email].timestamp < entry.timestamp)
                .forEach((entry: FeedbackSessionLogEntry) => this.studentToLog[entry.studentData.email] = entry);

            this.searchResult = this.toFeedbackSessionLogModel(targetFeedbackSessionLog);
          });
    }, (e: ErrorMessageOutput) => {
      this.statusMessageService.showErrorToast(e.error.message);
    });
  }

  private openModal(): void {
    const modalContent: string = 'Published date of selected feedback session is more than 30 days ago. '
      + 'Only activities within the last 30 days will be shown.';
    const modalRef: NgbModalRef =
      this.simpleModalService.openConfirmationModal('Continue the operation?', SimpleModalType.WARNING, modalContent);
    modalRef.result.then(
      () => {
        const earliestSearchDate: Date =
          new Date(Date.now() - this.LOGS_RETENTION_PERIOD_IN_MILLISECONDS + this.ONE_MINUTE_IN_MILLISECONDS);
        this.logsDateFrom = {
          year: earliestSearchDate.getFullYear(),
          month: earliestSearchDate.getMonth() + 1,
          day: earliestSearchDate.getDate(),
        };
        this.logsTimeFrom = {
          hour: earliestSearchDate.getHours(),
          minute: earliestSearchDate.getMinutes(),
        };

        this.notViewedSince = earliestSearchDate.getTime();
        this.search();
      },
      () => { this.isSearching = false; },
    );
  }

  private toFeedbackSessionLogModel(log: FeedbackSessionLog): FeedbackSessionLogModel {
    return {
      courseId: this.formModel.courseId,
      feedbackSessionName: this.formModel.feedbackSessionName,
      publishedDate: this.timezoneService.formatToString(
          this.publishedTime, log.feedbackSessionData.timeZone, this.LOGS_DATE_TIME_FORMAT),
      logColumnsData: [
        { header: 'Status', sortBy: SortBy.RESULT_VIEW_STATUS },
        { header: 'Name', sortBy: SortBy.GIVER_NAME },
        { header: 'Email', sortBy: SortBy.RESPONDENT_EMAIL },
        { header: 'Section', sortBy: SortBy.SECTION_NAME },
        { header: 'Team', sortBy: SortBy.TEAM_NAME },
      ],
      logRowsData: this.students
        .map((student: Student) => {
          let status: string;
          let dataStyle: string = 'font-family:monospace; white-space:pre;';
          if (student.email in this.studentToLog) {
            const entry: FeedbackSessionLogEntry = this.studentToLog[student.email];
            status = `Viewed last at   ${this.timezoneService.formatToString(entry.timestamp, log.feedbackSessionData.timeZone, this.LOGS_DATE_TIME_FORMAT)}`;
          } else {
            status = `Not viewed since ${this.timezoneService.formatToString(this.notViewedSince, log.feedbackSessionData.timeZone, this.LOGS_DATE_TIME_FORMAT)}`;
            dataStyle += 'color:red;';
          }
          return [
            { value: status,
              style: dataStyle },
            { value: student.name },
            { value: student.email },
            { value: student.sectionName },
            { value: student.teamName },
          ];
        }),
    };
  }
}
