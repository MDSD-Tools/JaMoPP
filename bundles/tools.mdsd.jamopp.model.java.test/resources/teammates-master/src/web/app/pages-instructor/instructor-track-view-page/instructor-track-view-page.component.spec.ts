import { HttpClientTestingModule } from '@angular/common/http/testing';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { of } from 'rxjs';

import { CourseService } from '../../../services/course.service';
import { FeedbackSessionsService } from '../../../services/feedback-sessions.service';
import { LogService } from '../../../services/log.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { StudentService } from '../../../services/student.service';
import { TimezoneService } from '../../../services/timezone.service';
import {
  Course,
  FeedbackSession,
  FeedbackSessionLog,
  FeedbackSessionLogType,
  FeedbackSessionPublishStatus,
  FeedbackSessionSubmissionStatus,
  ResponseVisibleSetting,
  SessionVisibleSetting,
  Student,
} from '../../../types/api-output';
import { SortBy } from '../../../types/sort-properties';
import { ColumnData } from '../../components/sortable-table/sortable-table.component';

import { InstructorTrackViewPageComponent } from './instructor-track-view-page.component';
import { InstructorTrackViewPageModule } from './instructor-track-view-page.module';
import Spy = jasmine.Spy;

describe('InstructorTrackViewPageComponent', () => {
  let component: InstructorTrackViewPageComponent;
  let fixture: ComponentFixture<InstructorTrackViewPageComponent>;
  let courseService: CourseService;
  let feedbackSessionsService: FeedbackSessionsService;
  let studentService: StudentService;
  let statusMessageService: StatusMessageService;
  let timezoneService: TimezoneService;
  let logService: LogService;

  const resultColumns: ColumnData[] = [
    { header: 'Status', sortBy: SortBy.RESULT_VIEW_STATUS },
    { header: 'Name', sortBy: SortBy.GIVER_NAME },
    { header: 'Email', sortBy: SortBy.RESPONDENT_EMAIL },
    { header: 'Section', sortBy: SortBy.SECTION_NAME },
    { header: 'Team', sortBy: SortBy.TEAM_NAME },
  ];

  const testCourse1: Course = {
    courseId: 'CS1234',
    courseName: 'test-course1',
    institute: 'Test Institute',
    timeZone: 'Asia/Singapore',
    creationTimestamp: 0,
    deletionTimestamp: 0,
  };

  const testCourse2: Course = {
    courseId: 'MA1234',
    courseName: 'test-course2',
    institute: 'Test Institute',
    timeZone: 'Asia/Singapore',
    creationTimestamp: 0,
    deletionTimestamp: 0,
  };

  const testFeedbackSessionPublished: FeedbackSession = {
    feedbackSessionName: 'First Session',
    courseId: 'CS1234',
    timeZone: 'Asia/Singapore',
    instructions: '',
    submissionStartTimestamp: 0,
    submissionEndTimestamp: 1549095330000,
    resultVisibleFromTimestamp: Date.now(),
    gracePeriod: 0,
    sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
    responseVisibleSetting: ResponseVisibleSetting.AT_VISIBLE,
    submissionStatus: FeedbackSessionSubmissionStatus.OPEN,
    publishStatus: FeedbackSessionPublishStatus.PUBLISHED,
    isClosingEmailEnabled: true,
    isPublishedEmailEnabled: true,
    createdAtTimestamp: 0,
  };

  const testFeedbackSessionNotPublished: FeedbackSession = {
    feedbackSessionName: 'Second Session',
    courseId: 'MA1234',
    timeZone: 'Asia/Singapore',
    instructions: '',
    submissionStartTimestamp: 0,
    submissionEndTimestamp: 1549095330000,
    gracePeriod: 0,
    sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
    responseVisibleSetting: ResponseVisibleSetting.AT_VISIBLE,
    submissionStatus: FeedbackSessionSubmissionStatus.OPEN,
    publishStatus: FeedbackSessionPublishStatus.NOT_PUBLISHED,
    isClosingEmailEnabled: true,
    isPublishedEmailEnabled: false,
    createdAtTimestamp: 0,
  };

  const testStudent: Student = {
    email: 'doejohn@email.com',
    courseId: 'CS1234',
    name: 'Doe John',
    teamName: 'team 1',
    sectionName: 'section 1',
  };

  const testLogs: FeedbackSessionLog = {
    feedbackSessionData: testFeedbackSessionPublished,
    feedbackSessionLogEntries: [
      {
        studentData: testStudent,
        feedbackSessionLogType: FeedbackSessionLogType.VIEW_RESULT,
        timestamp: 0,
      },
      {
        studentData: testStudent,
        feedbackSessionLogType: FeedbackSessionLogType.VIEW_RESULT,
        timestamp: 1000,
      },
    ],
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        InstructorTrackViewPageModule,
        NgbModule,
        HttpClientTestingModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(InstructorTrackViewPageComponent);
    courseService = TestBed.inject(CourseService);
    feedbackSessionsService = TestBed.inject(FeedbackSessionsService);
    studentService = TestBed.inject(StudentService);
    statusMessageService = TestBed.inject(StatusMessageService);
    timezoneService = TestBed.inject(TimezoneService);
    logService = TestBed.inject(LogService);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when page is still loading', () => {
    component.isLoading = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when searching for details in search form', () => {
    component.courses = [testCourse1, testCourse2];
    component.formModel = {
      courseId: 'CS1234',
      feedbackSessionName: 'First Session',
    };
    component.courseToFeedbackSession = {
      CS1234: [testFeedbackSessionPublished],
      MA1234: [],
    };
    component.isLoading = false;
    component.isSearching = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should snap with results of a search', () => {
    component.searchResult = {
      courseId: 'CS1234',
      feedbackSessionName: 'First Session',
      publishedDate: 'Thu, 27 May, 2021 12:00:00 PM',
      logColumnsData: resultColumns,
      logRowsData: [[
        { value: 'Sun, 30 May, 2021 12:30:00 PM' },
        { value: 'Doe John' },
        { value: 'doejohn@email.com' },
        { value: 'section 1' },
        { value: 'team 1' },
      ]],
    };
    component.isLoading = false;
    component.isSearching = false;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should load all courses and feedback sessions that instructor has on init', () => {
    const courseSpy: Spy = spyOn(courseService, 'getAllCoursesAsInstructor').and
        .returnValue(of({ courses: [testCourse1, testCourse2] }));
    const feedbackSessionSpy: Spy = spyOn(feedbackSessionsService, 'getFeedbackSessionsForInstructor').and
        .returnValue(of({ feedbackSessions: [testFeedbackSessionPublished] }));

    component.ngOnInit();

    expect(component.isLoading).toBeFalsy();
    expect(courseSpy).toBeCalledWith('active');
    expect(component.courses.length).toEqual(2);
    expect(component.courseToFeedbackSession[testCourse1.courseId][0]).toEqual(testFeedbackSessionPublished);
    expect(feedbackSessionSpy).toHaveBeenNthCalledWith(1, testCourse1.courseId);
    expect(feedbackSessionSpy).toHaveBeenNthCalledWith(2, testCourse2.courseId);
  });

  it('should load students and search for logs when view button is clicked', () => {
    spyOn(feedbackSessionsService, 'getFeedbackSession').and.returnValue(of(testFeedbackSessionPublished));
    spyOn(studentService, 'getStudentsFromCourse').and.returnValue(of({ students: [testStudent] }));
    const logSpy: Spy = spyOn(logService, 'searchFeedbackSessionLog').and
        .returnValue(of({ feedbackSessionLogs: [testLogs] }));
    const timezoneSpy: Spy = spyOn(timezoneService, 'resolveLocalDateTime').and
        .returnValue(0);

    component.isLoading = false;
    component.isSearching = false;
    component.students = [];
    component.formModel = {
      courseId: testCourse1.courseId,
      feedbackSessionName: testFeedbackSessionPublished.feedbackSessionName,
    };
    component.courseToFeedbackSession = { CS1234: [testFeedbackSessionPublished] };
    fixture.detectChanges();

    fixture.debugElement.nativeElement.querySelector('#view-button').click();

    expect(timezoneSpy).toHaveBeenCalledTimes(2);
    expect(component.students.length).toEqual(1);
    expect(component.students[0]).toEqual(testStudent);
    expect(logSpy).toHaveBeenCalled();
    expect(logSpy).toHaveBeenCalledWith({
      courseId: testCourse1.courseId,
      searchFrom: '0',
      searchUntil: '0',
      sessionName: testFeedbackSessionPublished.feedbackSessionName,
    });
  });

  it('should display error message if feedback session is not published', () => {
    spyOn(feedbackSessionsService, 'getFeedbackSession').and.returnValue(of(testFeedbackSessionNotPublished));
    const messageSpy: Spy = spyOn(statusMessageService, 'showErrorToast');

    component.isLoading = false;
    component.isSearching = false;
    component.students = [];
    component.formModel = {
      courseId: testCourse2.courseId,
      feedbackSessionName: testFeedbackSessionNotPublished.feedbackSessionName,
    };
    component.courseToFeedbackSession = { MA1234: [testFeedbackSessionNotPublished] };
    fixture.detectChanges();

    fixture.debugElement.nativeElement.querySelector('#view-button').click();

    expect(messageSpy).lastCalledWith('This feedback session is not published');
  });
});
