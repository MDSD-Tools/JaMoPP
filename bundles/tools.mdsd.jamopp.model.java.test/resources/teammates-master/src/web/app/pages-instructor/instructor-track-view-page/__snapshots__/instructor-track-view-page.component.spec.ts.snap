// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstructorTrackViewPageComponent should snap when page is still loading 1`] = `
<tm-instructor-track-view-page
  LOGS_DATE_TIME_FORMAT={[Function String]}
  LOGS_RETENTION_PERIOD_IN_DAYS={[Function Number]}
  LOGS_RETENTION_PERIOD_IN_MILLISECONDS={[Function Number]}
  ONE_MINUTE_IN_MILLISECONDS={[Function Number]}
  SortBy={[Function Object]}
  courseService={[Function CourseService]}
  courseToFeedbackSession={[Function Object]}
  courses={[Function Array]}
  feedbackSessionsService={[Function FeedbackSessionsService]}
  formModel={[Function Object]}
  hasResult="false"
  isLoading={[Function Boolean]}
  isSearching="false"
  logsDateFrom={[Function Object]}
  logsDateTo={[Function Object]}
  logsService={[Function LogService]}
  logsTimeFrom={[Function Object]}
  logsTimeTo={[Function Object]}
  notViewedSince="0"
  publishedTime="0"
  searchResult={[Function Object]}
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
  studentToLog={[Function Object]}
  students={[Function Array]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Student Result View Tracking
  </h1><p>
    This page allows you to track if the session results were viewed by students. The search will go only up to the date the session was published. The earliest date you can search for is 
    <b>
      30 days
    </b>
     before today.
  </p><hr /><tm-loading-spinner>
    <div
      class="loading-container"
    >
      <div
        class="row spinner-border text-primary"
        role="status"
      />
      <div
        class="row text-secondary"
      >
        Loading...
      </div>
    </div>
  </tm-loading-spinner><div>
    
  </div>
</tm-instructor-track-view-page>
`;

exports[`InstructorTrackViewPageComponent should snap when searching for details in search form 1`] = `
<tm-instructor-track-view-page
  LOGS_DATE_TIME_FORMAT={[Function String]}
  LOGS_RETENTION_PERIOD_IN_DAYS={[Function Number]}
  LOGS_RETENTION_PERIOD_IN_MILLISECONDS={[Function Number]}
  ONE_MINUTE_IN_MILLISECONDS={[Function Number]}
  SortBy={[Function Object]}
  courseService={[Function CourseService]}
  courseToFeedbackSession={[Function Object]}
  courses={[Function Array]}
  feedbackSessionsService={[Function FeedbackSessionsService]}
  formModel={[Function Object]}
  hasResult="false"
  isLoading="false"
  isSearching={[Function Boolean]}
  logsDateFrom={[Function Object]}
  logsDateTo={[Function Object]}
  logsService={[Function LogService]}
  logsTimeFrom={[Function Object]}
  logsTimeTo={[Function Object]}
  notViewedSince="0"
  publishedTime="0"
  searchResult={[Function Object]}
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
  studentToLog={[Function Object]}
  students={[Function Array]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Student Result View Tracking
  </h1><p>
    This page allows you to track if the session results were viewed by students. The search will go only up to the date the session was published. The earliest date you can search for is 
    <b>
      30 days
    </b>
     before today.
  </p><hr /><div
    class="card bg-form no-border"
  >
    <div
      class="card-body"
    >
      <div
        class="row form-group"
      >
        <div
          class="col-5"
        >
          <div
            class="col-12"
          >
            <label
              class="font-weight-bold"
              for="course-id-dropdown"
            >
              Course ID
            </label>
          </div>
          <div
            class="col-12"
          >
            <select
              class="form-control col-12 ng-untouched ng-pristine ng-valid"
              id="course-id-dropdown"
            >
              <option
                disabled=""
                hidden=""
                selected=""
                value=""
              >
                Select Course ID
              </option>
              
              <option>
                 CS1234 
              </option>
              <option>
                 MA1234 
              </option>
            </select>
          </div>
        </div>
        <div
          class="col-5"
        >
          <div
            class="col-12"
          >
            <label
              class="font-weight-bold"
              for="session-dropdown"
            >
              Feedback Session Name
            </label>
          </div>
          <div
            class="col-12"
          >
            <select
              class="form-control col-md-12 ng-untouched ng-pristine ng-valid"
              id="session-dropdown"
            >
              
              <option
                value="First Session"
              >
                 First Session 
              </option>
            </select>
          </div>
        </div>
        <div
          class="col-1 mt-4"
        >
          <button
            class="float-right btn btn-primary"
            disabled=""
            id="view-button"
          >
            View
          </button>
        </div>
      </div>
    </div>
  </div><tm-loading-spinner>
    <div
      class="loading-container"
    >
      <div
        class="row spinner-border text-primary"
        role="status"
      />
      <div
        class="row text-secondary"
      >
        Loading...
      </div>
    </div>
  </tm-loading-spinner>
</tm-instructor-track-view-page>
`;

exports[`InstructorTrackViewPageComponent should snap with default fields 1`] = `
<tm-instructor-track-view-page
  LOGS_DATE_TIME_FORMAT={[Function String]}
  LOGS_RETENTION_PERIOD_IN_DAYS={[Function Number]}
  LOGS_RETENTION_PERIOD_IN_MILLISECONDS={[Function Number]}
  ONE_MINUTE_IN_MILLISECONDS={[Function Number]}
  SortBy={[Function Object]}
  courseService={[Function CourseService]}
  courseToFeedbackSession={[Function Object]}
  courses={[Function Array]}
  feedbackSessionsService={[Function FeedbackSessionsService]}
  formModel={[Function Object]}
  hasResult="false"
  isLoading={[Function Boolean]}
  isSearching="false"
  logsDateFrom={[Function Object]}
  logsDateTo={[Function Object]}
  logsService={[Function LogService]}
  logsTimeFrom={[Function Object]}
  logsTimeTo={[Function Object]}
  notViewedSince="0"
  publishedTime="0"
  searchResult={[Function Object]}
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
  studentToLog={[Function Object]}
  students={[Function Array]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Student Result View Tracking
  </h1><p>
    This page allows you to track if the session results were viewed by students. The search will go only up to the date the session was published. The earliest date you can search for is 
    <b>
      30 days
    </b>
     before today.
  </p><hr /><tm-loading-spinner>
    <div
      class="loading-container"
    >
      <div
        class="row spinner-border text-primary"
        role="status"
      />
      <div
        class="row text-secondary"
      >
        Loading...
      </div>
    </div>
  </tm-loading-spinner><div>
    
  </div>
</tm-instructor-track-view-page>
`;

exports[`InstructorTrackViewPageComponent should snap with results of a search 1`] = `
<tm-instructor-track-view-page
  LOGS_DATE_TIME_FORMAT={[Function String]}
  LOGS_RETENTION_PERIOD_IN_DAYS={[Function Number]}
  LOGS_RETENTION_PERIOD_IN_MILLISECONDS={[Function Number]}
  ONE_MINUTE_IN_MILLISECONDS={[Function Number]}
  SortBy={[Function Object]}
  courseService={[Function CourseService]}
  courseToFeedbackSession={[Function Object]}
  courses={[Function Array]}
  feedbackSessionsService={[Function FeedbackSessionsService]}
  formModel={[Function Object]}
  hasResult="false"
  isLoading="false"
  isSearching="false"
  logsDateFrom={[Function Object]}
  logsDateTo={[Function Object]}
  logsService={[Function LogService]}
  logsTimeFrom={[Function Object]}
  logsTimeTo={[Function Object]}
  notViewedSince="0"
  publishedTime="0"
  searchResult={[Function Object]}
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
  studentToLog={[Function Object]}
  students={[Function Array]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Student Result View Tracking
  </h1><p>
    This page allows you to track if the session results were viewed by students. The search will go only up to the date the session was published. The earliest date you can search for is 
    <b>
      30 days
    </b>
     before today.
  </p><hr /><div
    class="card bg-form no-border"
  >
    <div
      class="card-body"
    >
      <div
        class="row form-group"
      >
        <div
          class="col-5"
        >
          <div
            class="col-12"
          >
            <label
              class="font-weight-bold"
              for="course-id-dropdown"
            >
              Course ID
            </label>
          </div>
          <div
            class="col-12"
          >
            <select
              class="form-control col-12 ng-untouched ng-pristine ng-valid"
              id="course-id-dropdown"
            >
              <option
                disabled=""
                hidden=""
                selected=""
                value=""
              >
                Select Course ID
              </option>
              
            </select>
          </div>
        </div>
        <div
          class="col-5"
        >
          <div
            class="col-12"
          >
            <label
              class="font-weight-bold"
              for="session-dropdown"
            >
              Feedback Session Name
            </label>
          </div>
          <div
            class="col-12"
          >
            <select
              class="form-control col-md-12 ng-untouched ng-pristine ng-valid"
              id="session-dropdown"
            >
              
            </select>
          </div>
        </div>
        <div
          class="col-1 mt-4"
        >
          <button
            class="float-right btn btn-primary"
            disabled=""
            id="view-button"
          >
            View
          </button>
        </div>
      </div>
    </div>
  </div><div>
    
  </div>
</tm-instructor-track-view-page>
`;
